core/node/node_framework/src/service/mod.rs

```diff
@@ -82,6 +90,22 @@ impl ZkStackService {
         self
     }
 
+    /// Setups a hook that will be called before the service starts running the tasks.
+    ///
+    /// This hook will be invoked after all the wiring layers have been processed and all the tasks
+    /// have been added to the service, but before any of the tasks are started. During the setup, the
+    /// hook can launch any added tasks and access resources through the [`PreRun`] object.
+    ///
+    /// The node will block on the hook execution if its provided, so if it's a long-running operation,
+    /// it is advised to start the supplementary tasks, such as healthcheck server, in the hook itself.
+    pub fn run_with_setup<F>(mut self, setup: F) -> anyhow::Result<()>
+    where
+        F: FnOnce(&mut PreRun) -> BoxFuture<anyhow::Result<()>> + Send + 'static,
```

Comment:
Since the method has type params anyway, is there anything preventing replacing `BoxFuture` with another type param? Would make calling this method more eloquent.

Reply:
`BoxFuture` is used within the `SetupHook` type, and I'm not sure how to perform a conversion from `dyn FnOnce(..) -> Fut` (where `Fut: Future<..>`) to `dyn FnOnce(..) -> BoxFuture<..>` without wrapping it into an additional closure.
Could you please give me a hint here?

Reply:
You're right, it doesn't look possible; it'd require async closures or syntax like

```rust
pub fn run_with_setup<F, Fut>(mut self, setup: F) -> anyhow::Result<()>
where
    F: for<'a> FnOnce(&'a mut PreRun) -> Fut<'a> + Send + 'static,
    for<'a> Fut<'a>: Future<Output = anyhow::Result<()>> + Send + 'a,
```